lib/util.js

```diff
@@ -361,261 +371,202 @@ function formatValue(ctx, value, recurseTimes) {
-
-  // Make dates with properties first say the date
-  if (isDate(value)) {
-    base = ` ${dateToISOString.call(value)}`;
-  }
-
-  // Make error with message first say the error
-  if (isError(value)) {
-    base = ` ${formatError(value)}`;
-  }
+    var raw;
+    // This could be a boxed primitive (new String(), etc.), check valueOf()
+    // NOTE: Avoid calling `valueOf` on `Date` instance because it will return
+    // a number which, when object has some additional user-stored `keys`,
+    // will be printed out.
+    // The .valueOf() call can fail for a multitude of reasons
+    try {
+      raw = value.valueOf();
+    } catch (e) { /* ignore */ }
 
-  // Make boxed primitive Strings look like such
-  if (typeof raw === 'string') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [String: ${formatted}]`;
+    if (typeof raw === 'string') {
+      const formatted = formatPrimitiveNoColor(ctx, raw);
+      if (keyLength === raw.length) {
+        return ctx.stylize(`[String: ${formatted}]`, 'string');
+      }
+      base = ` [String: ${formatted}]`;
+      // For boxed Strings, we have to remove the 0-n indexed entries,
+      // since they just noisy up the output and are redundant
+      // Make boxed primitive Strings look like such
+      keys = keys.slice(value.length);
+    } else if (typeof raw === 'number') {
+      // Make boxed primitive Numbers look like such
+      const formatted = formatPrimitiveNoColor(ctx, raw);
+      if (keyLength === 0)
+        return ctx.stylize(`[Number: ${formatted}]`, 'number');
+      base = ` [Number: ${formatted}]`;
+    } else if (typeof raw === 'boolean') {
+      // Make boxed primitive Booleans look like such
+      const formatted = formatPrimitiveNoColor(ctx, raw);
+      if (keyLength === 0)
+        return ctx.stylize(`[Boolean: ${formatted}]`, 'boolean');
+      base = ` [Boolean: ${formatted}]`;
+    } else if (typeof raw === 'symbol') {
+      const formatted = formatPrimitiveNoColor(ctx, raw);
+      return ctx.stylize(`[Symbol: ${formatted}]`, 'symbol');
+    } else if (keyLength === 0) {
+      if (isExternal(value))
+        return ctx.stylize('[External]', 'special');
+      return `${ctorName}${braces[0]}${base}${braces[1]}`;
+    } else {
+      braces[0] = `${ctorName}{`;
+    }
   }
 
-  // Make boxed primitive Numbers look like such
-  if (typeof raw === 'number') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [Number: ${formatted}]`;
+  if (recurseTimes !== null) {
+    if (recurseTimes < 0)
+      return ctx.stylize('[Object]', 'special');
+    recurseTimes -= 1;
   }
 
-  // Make boxed primitive Booleans look like such
-  if (typeof raw === 'boolean') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [Boolean: ${formatted}]`;
-  }
+  // Using an array here is actually better for the average case then using
+  // a Set. `seen` will only check for the depth and will never grow to large.
+  if (ctx.seen.includes(value))
```

Comment:
I reverted my change to a Set here: the `seen` part will only contain up to `ctx.depth + 1` elements and in the regular case that is three. Using a Set for such a low amount of elements seems unnecessary, as it increases the code complexity due to having to handle both, an array and a set in case customInspect is set.